File: Data Structures/Graphs/Path Finding/FloydWarshall.cpp
#include<iostream>
#include<vector>
#include<math.h>
#include<iomanip>
#include<algorithm>
#include<string>
#include<map>
#include<bitset>
#include<cmath>
#include<iomanip>
#define vint vector<int>
#define ll long long
#define mod 1000000007
#define vll vector<long long>
#define ld long double
#define mii map<int,int>
#define miv map<int,vector<int>>
#define vf vector<float>
#define vd vector<double>
#define vii vector<vector<int>>

/*

Floyd Warshall's Algorithm

Time Complexity : O(N^3) ,  N -> Number of vertices
Space Complexity : O(N^2) ,  N -> Number of vertices

*/

using namespace std;

void solve()
{
	// Input
	int n,e;
	cin>>n>>e;
	vii distance;
	for(int i=0;i<n;i++)
	{
		vint r;
		for(int j=0;j<n;j++)
		{
			r.push_back(mod);
		}
		distance.push_back(r);
	}

	// Initialising Distance Matrix
	for(int i=0;i<e;i++)
	{
		int a,b,w;
		cin>>a>>b>>w;
		a--;
		b--;
		distance[a][b]=w;
	}

	// Floyd's Algorithm
	for(int k=0;k<n;k++)
	{
		for(int i=0;i<n;i++)
		{
			for(int j=0;j<n;j++)
			{
				if(distance[i][k]+distance[k][j]<distance[i][j])
				{
					distance[i][j]=distance[i][k]+distance[k][j];
				}
			}
		}
	}

	// Printing Result
	for(int i=0;i<n;i++)
	{
		for(int j=0;j<n;j++)
		{
			if(distance[i][j]==mod)
			{
				distance[i][j]=-1;
			}
			cout<<distance[i][j]<<"\t";
		}
		cout<<endl;
	}
	return;
}

int main()
{
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);

	#ifndef ONLINE_JUDGE
	freopen("input.txt","r",stdin);
	freopen("output.txt","w",stdout);
	#endif

	int t=1;
	while(t--)
	{
		solve();
	}
	return 0;
}

File: Algorithms/Dynamic Programming/FloydWarshall.py
import math
import sys
import collections
import random
input=sys.stdin.readline
print=sys.stdout.write
sys.setrecursionlimit(1000000)
mod=1000000007

'''

Floyd Warshall's Algorithm

Time Complexity : O(N^3) ,  N -> Number of vertices
Space Complexity : O(N^2) ,  N -> Number of vertices

'''

def lin():
    return list(map(int,input().split()))

def solve():
    # Input
    n,e=lin()
    distance=[]
    
    # Initialising Distance Matrix
    for _ in range(n):
        r=[mod]*n
        distance.append(r)
    
    for _ in range(e):
        a,b,w=lin()
        distance[a-1][b-1]=w
    
    # Floyd's Algorithm
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if distance[i][k]+distance[k][j]<distance[i][j]:
                    distance[i][j]=distance[i][k]+distance[k][j]
    
    # Printing Result
    for i in range(n):
        for j in range(n):
            if distance[i][j]==mod:
                distance[i][j]=-1
            print(str(distance[i][j])+"\t")
        print("\n")
    return

if __name__=='__main__':
    t=1
    while(t):
        t-=1
        solve()
